contributors/design-proposals/rules-review-api.md

```diff
@@ -0,0 +1,126 @@
+
+[kubernetes/kubernetes#48051](https://github.com/kubernetes/kubernetes/pull/48051) implements most of this proposal.
+
+## API additions
+
+Add a top level type to the `authorization.k8s.io` API group called `SelfSubjectRulesReview`. This mirrors the existing `SelfSubjectAccessReview`.
+
+```
+type SelfSubjectRulesReview struct {
+	metav1.TypeMeta
+
+	Spec SelfSubjectRulesReviewSpec
+
+	// Status is filled in by the server and represents the set of actions a user can perform.
+	Status SubjectRulesReviewStatus
+}
+
+type SelfSubjectRulesReviewSpec struct {
+	// Namespace to evaluate rules for. Required.
+	Namespace string
+}
+
+type SubjectRulesReviewStatus struct {
+	// ResourceRules is the list of actions the subject is allowed to perform on resources.
+	// The list ordering isn't significant, may contain duplicates, and possibly be incomplete.
+	ResourceRules []ResourceRule
+	// NonResourceRules is the list of actions the subject is allowed to perform on non-resources.
+	// The list ordering isn't significant, may contain duplicates, and possibly be incomplete.
+	NonResourceRules []NonResourceRule
+	// EvaluationError can appear in combination with Rules. It indicates an error occurred during
+	// rule evaluation, such as an authorizer that doesn't support rule evaluation, and that
+	// ResourceRules and/or NonResourceRules may be incomplete.
+	EvaluationError string
+}
+```
+
+The `ResourceRules` and `NonResourceRules` rules are similar to the types use by RBAC and the internal authorization system.
+
+```
+# docstrings omitted for brevity. 
+type ResourceRule struct {
+	Verbs []string
+	APIGroups []string
+	Resources []string
+	ResourceNames []string
+}
+
+type NonResourceRule struct {
+	Verbs []string
+	NonResourceURLs []string
+}
+```
+
+All of these fields can include the string `*` to indicate all values are allowed.
+
+### Differences from OpenShift: user extras vs. scopes
+
+OpenShift `SelfSubjectRulesReviewSpec` takes a set of [`Scopes`][openshift-scopes]. This lets OpenShift clients use the API for queries such as _"what could I do if I provide this scope to limit my credentials?"_
+
+ In core kube, scopes are replaced by "user extras" field, a map of opaque strings that can be used for implementation specific user data. Unlike OpenShift, where scopes are always used to restrict credential powers, user extras are commonly used to expand powers. For example, the proposed [Keystone authentiator][keystone-authn] used them to include additional roles and project fields.
+
+Since user extras can be used to expand the power of users, instead of only restricting, this proposal argues that `SelfSubjectRulesReview` shouldn't let a client specify them like `Scopes`. It wouldn't be within the spirit of a `SelfSubject` resource to let a user determine information about other projects or roles.
```

Comment:
> Extra fields should NOT be used to grant access to new namespaces. Only to modify the
> access of a user within a namespace.

Granting access to other namespaces does not leak information, since if I lie about my extra and see "yes you have access to namespace X", I'd get that result regardless of whether the namespace exists.